algorithm-simulator/src/main/scala/style95/QueueSimulator.scala

```diff
@@ -52,13 +63,27 @@ class QueueSimulator(scaler: Scaler,
       creating -= sender
       existing += sender -> ContainerStatus(true)
       tryRunActions()
+    case ContainerStopped =>
+      if (!deleting.contains(sender)) {
+        log.error(
+          s"state transition violation: $sender is not in deleting state")
+      }
+      deleting -= sender
     case WorkDone(msg @ ActivationMessage(requester, start, invoked)) =>
+      val invokeTime = System.nanoTime() - invoked
       outSinceLastTick += 1
-      existing += sender -> ContainerStatus(true)
+      // avoid scheduling action to a deleted container
+      if (!deleting.contains(sender)) {
+        existing += sender -> ContainerStatus(true)
+      }
       requester ! msg
-      logger ! ActivationRecord(elapsed,
-                                invoked - start,
-                                System.nanoTime() - invoked)
+
+      // incremental average
+      completedActivations += 1
+      averageInvokeTime += 1.0 / completedActivations * (invokeTime - averageInvokeTime)
+
+      logger ! ActivationRecord(elapsed, invoked - start, invokeTime)
+
       tryRunActions()
```

Comment:
I think I should comment on this previously sorry for this but just came up in mind.
Actually, it's not easy to manage the status of idle containers in the real scene.

This is the reason why I proposed a pull-based model in my new architecture.
https://cwiki.apache.org/confluence/display/OPENWHISK/Design+consideration

I am surely fine with this version but if this simulator changes its execution model to the pull-based one, it would better emulate the real environment.

Reply:
Indeed, this version of simulation is not purely polling based. It should be changed. However, I cannot come up with a better implementation. Could you please drop me some enlightenment?

Reply:
I think we can handle that in the subsequent PR.

Reply:
In my implementation, I created a queue which keeps the incoming activations and once containers are created for the given action, they autonomously request the activation and proceed them.